proto/server/datastore/datastore.proto

```diff
@@ -2,348 +2,329 @@ syntax = "proto3";
 }
 
-// Represents a list of Node resolver map entries for the specified Spiffe ID
-message FetchNodeResolverMapEntryResponse {
-    // List of Node resolver map entries
-    repeated NodeResolverMapEntry nodeResolverMapEntryList = 1;
+message FetchAttestedNodeEntryResponse {
+    AttestedNodeEntry entry = 1;
 }
 
+message ListAttestedNodeEntriesRequest {
+    google.protobuf.Int64Value by_expires_before = 1;
+}
 
-// Represents a Node resolver map entry to delete
-message DeleteNodeResolverMapEntryRequest {
-    // Node resolver map entry
-    NodeResolverMapEntry nodeResolverMapEntry = 1;
+message ListAttestedNodeEntriesResponse {
+    repeated AttestedNodeEntry entries = 1;
 }
 
-// Represents a list of Node resolver map entries
-message DeleteNodeResolverMapEntryResponse {
-    // List of Node resolver map entries
-    repeated NodeResolverMapEntry nodeResolverMapEntryList = 1;
+message UpdateAttestedNodeEntryRequest {
+    string spiffe_id = 1;
+
+    string cert_serial_number = 2;
+
+    int64 cert_not_after = 3;
 }
 
+message UpdateAttestedNodeEntryResponse {
+    AttestedNodeEntry entry = 1;
+}
 
-// Represents a list of Node resolver map entries
-message RectifyNodeResolverMapEntriesRequest {
-    // List of Node resolver map entries
-    repeated NodeResolverMapEntry nodeResolverMapEntryList = 1;
+message DeleteAttestedNodeEntryRequest {
+    string spiffe_id = 1;
 }
 
-// Represents a list of Node resolver map entries
-message RectifyNodeResolverMapEntriesResponse {
-    // List of Node resolver map entries
-    repeated NodeResolverMapEntry nodeResolverMapEntryList = 1;
+message DeleteAttestedNodeEntryResponse {
+    AttestedNodeEntry entry = 1;
 }
 
-//
-//
-//
 
-// Represents a Registration entry to create
+/////////////////////////////////////////////////////////////////////////////
+// Registration Entries
+/////////////////////////////////////////////////////////////////////////////
+
 message CreateRegistrationEntryRequest {
-    // Registration entry
-    spire.common.RegistrationEntry registeredEntry = 1;
+    spire.common.RegistrationEntry entry = 1;
 }
 
-// Represents the created Registration entry
 message CreateRegistrationEntryResponse {
-    // Registration entry ID
-    string registeredEntryId = 1;
+    string entry_id = 1;
 }
 
-
-// Represents a Registration entry ID to fetch
 message FetchRegistrationEntryRequest {
-    // Registration entry ID
-    string registeredEntryId = 1;
+    string entry_id = 1;
 }
 
-// Represents a Registration entry
 message FetchRegistrationEntryResponse {
-    // Registration entry
-    spire.common.RegistrationEntry registeredEntry = 1;
+    spire.common.RegistrationEntry entry = 1;
 }
 
-// Represents a list of Registration entries
-message FetchRegistrationEntriesResponse {
-    // Registration entries
-    spire.common.RegistrationEntries registeredEntries = 1;
+message BySelectors {
+    repeated spire.common.Selector selectors = 1;
+    bool allow_any_combination = 2;
```

Comment:
Suggest `exact_match` here instead - I had to go review the source to figure out what this did

This boolean gives us either exact match or subset matches. The former is useful for duplicate checks, and the latter is useful for attestation. I wonder if we also will want superset matches, which could be useful in cases where the operator doesn't know the full set of selectors e.g. "show me all entries referencing this particular docker image"

Reply:
I originally had it that way but I thought that it was a bit more dangerous having the default behavior (in case you forget to set the value) hand back more registration entries than intended...

Reply:
How about `include_subsets` then or something similar?

Reply:
I guess the confusion for me is that the term `combination` makes me think that order matters

Reply:
Awesome. Let's do an enum type, with EXACT_MATCH as the default. That allows us to handle the superset condition later, if/when it becomes useful.